app/models/spree/gateway/mollie_gateway.rb

```diff
@@ -215,23 +222,36 @@ def update_payment_status(payment)
           api_key: get_preference(:api_key)
       )
 
-      MollieLogger.debug("Updating order state for payment. Payment has state #{mollie_payment.status}")
-
+      MollieLogger.debug("Checking Mollie payment status. Mollie payment has status #{mollie_payment.status}")
       update_by_mollie_status!(mollie_payment, payment)
     end
 
     def update_by_mollie_status!(mollie_payment, payment)
       case mollie_payment.status
-        when 'paid'
-          payment.complete! unless payment.completed?
+      when 'paid'
+        # If Mollie payment is already paid and refunded amount is more than 0, don't update payment
+        if mollie_payment.paid? && mollie_payment.amount_refunded.value > 0
+          MollieLogger.debug('Payment is refunded. Not updating the payment status within Spree.')
+          return
+        end
+
+        if payment.completed?
+          MollieLogger.debug('Payment is already completed. Not updating the payment status within Spree.')
+        end
+
+        # If order is already paid for, don't mark it as complete again.
+        unless payment.completed?
+          payment.complete!
           payment.order.finalize!
           payment.order.update_attributes(:state => 'complete', :completed_at => Time.now)
-        when 'canceled', 'expired', 'failed'
-          payment.failure! unless payment.failed?
-          payment.order.update_attributes(:state => 'payment', :completed_at => nil)
-        else
-          MollieLogger.debug('Unhandled Mollie payment state received. Therefore we did not update the payment state.')
-          payment.order.update_attributes(state: 'payment', completed_at: nil)
+          MollieLogger.debug('Payment is paid and will transition to completed. Order will be finalized.')
+        end
+      when 'canceled', 'expired', 'failed'
+        payment.failure! unless payment.failed?
+        payment.order.update_attributes(:state => 'payment', :completed_at => nil)
+      else
+        MollieLogger.debug('Unhandled Mollie payment state received. Therefore we did not update the payment state.')
+        payment.order.update_attributes(state: 'payment', completed_at: nil)
```

Comment:
We're mixing "Ruby <1.9 " style hash notation with "Ruby>=1.9" style hash notation. If we don't aim to support ruby 1.8 (which was EOLled June 2013) we should migrate to Ruby 1.9 style hash notation. Especially since _some_ 1.9 style notation is in here, that would hurt nobody that's already using this sucesfully.

Reply:
I think we should switch to "Ruby>=1.9" style hash notation. We cannot support Ruby 1.8, since we depend on `mollie-api-ruby` (which requires "Ruby>=2.1"). I'll create an issue to migrate the entire codebase 👍

Reply:
Created #32. Will pick that up later.